### PR TITLE
Problem: refactor topology services - PQSWMBT-3496

### PR DESCRIPTION
**Pull request by jeffdegott**
Solution: move topology code in fty-asset, branch REST api's on fty-asset mailbox

### CODE REVIEW

File: src/web/src/topology_location.ecpp [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+#include <fty_common_db.h>
+#include <fty_common_mlm_tntmlm.h>
+
+// set S with MSG popped frame (no memleak, S untouched if NULL frame)
+static void zmsg_pop_s (zmsg_t *msg, std::string & s)
+{
+    char *aux = msg ? zmsg_popstr (msg) : NULL;
+    if (aux) s = aux;
+    zstr_free (&aux);
+}
```
Comment (eric3873) [inline src/web/src/topology_location.ecpp]: Possible to factorize if common section available

File: src/web/src/topology_location.ecpp [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+            asset_id = from;
+        }
+        else if (!to.empty()) {
+            parameter_name = "to";
+            asset_id = to;
+        }
+        else {
+            log_error ("unexpected parameter");
+            std::string err = TRANSLATE_ME("Unexpected parameter");
+            http_die ("internal-error", err.c_str());
```
Comment (eric3873) [inline src/web/src/topology_location.ecpp]: Dead code

File: src/web/src/topology_location.ecpp [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+                //    http_die("request-param-bad", "feed_by", feed_by.c_str (), expected.c_str ());
+                //}
+            }
+
+            // set options (json payload)
+            options = "{ ";
+            options.append("\"filter\": \"").append(filter).append("\", ");
+            options.append("\"feed_by\": \"").append(feed_by).append("\", ");
+            options.append("\"recursive\": ").append(recursive); // bool
+            options.append(" }");
```
Comment (eric3873) [inline src/web/src/topology_location.ecpp]: Expression with << more lisible

File: src/web/src/topology_location.ecpp [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    zmsg_addstr (req, asset_id.c_str ());
+    if (!options.empty()) zmsg_addstr (req, options.c_str ());
+    zmsg_t *resp = client->requestreply (ADDRESS, SUBJECT, 5, &req);
+    zmsg_destroy (&req);
+
+    #define CLEANUP { zmsg_destroy (&resp); }
+
+    if (!resp) {
+        CLEANUP;
+        log_error ("client->requestreply (timeout = '5') failed");
```
Comment (eric3873) [inline src/web/src/topology_location.ecpp]: #define for timeout (use several times in file)

File: src/web/src/topology_input_power_chain.ecpp [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        std::string err = TRANSLATE_ME("Memory allocation failed.");
+        http_die ("internal-error", err.c_str());
+    }
+
+    zmsg_addstr (req, COMMAND);
+    zmsg_addstr (req, asset_id.c_str ());
+    zmsg_t *resp = client->requestreply (ADDRESS, SUBJECT, TIMEOUT, &req);
+    zmsg_destroy (&req);
+
+    #define CLEANUP { zmsg_destroy (&resp); }
```
Comment (boricj) [inline src/web/src/topology_input_power_chain.ecpp]: Consider using RAII-enabled ZmsgGuard instead.

File: src/web/src/topology_input_power_chain.ecpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        std::string err = TRANSLATE_ME("Connection to mlm client failed.");
+        http_die ("internal-error", err.c_str());
+    }
+
+    // set/send req, recv response
+    zmsg_t *req = zmsg_new ();
+    if (!req) {
+        log_error ("zmsg_new () failed");
+        std::string err = TRANSLATE_ME("Memory allocation failed.");
+        http_die ("internal-error", err.c_str());
```
Comment (boricj) [inline src/web/src/topology_input_power_chain.ecpp]: Inline string construction:
```c++
http_die ("internal-error", TRANSLATE_ME("Memory allocation failed.").c_str());
```

File: src/web/src/topology_input_power_chain.ecpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        std::string expected = TRANSLATE_ME("existing asset name");
+        http_die ("request-param-bad", "dc_id", asset_id.c_str (), expected.c_str ());
+    }
+
+    // db checks
+    {
+        // asset_id valid?
+        if (!persist::is_ok_name (asset_id.c_str ()) ) {
+            std::string expected = TRANSLATE_ME("valid asset name");
+            http_die ("request-param-bad", "dc_id", asset_id.c_str (), expected.c_str ());
```
Comment (boricj) [inline src/web/src/topology_input_power_chain.ecpp]: Inline string construction:
```c++
http_die ("request-param-bad", "dc_id", asset_id.c_str (), TRANSLATE_ME("valid asset name").c_str ());
```

File: src/web/src/topology_input_power_chain.ecpp [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    //ftylog_setVeboseMode(ftylog_getInstance());
+    log_trace ("in %s", request.getUrl().c_str ());
+
+    const char *ADDRESS = AGENT_FTY_ASSET; // "asset-agent" 42ty/fty-asset
+    const char *SUBJECT = "TOPOLOGY";
+    const char *COMMAND = "INPUT_POWERCHAIN";
+    const int TIMEOUT = 5; // seconds
+
+    // get param, id of datacenter retrieved from url
+    std::string asset_id = request.getArg ("id");
```
Comment (boricj) [inline src/web/src/topology_input_power_chain.ecpp]: Make `asset_id` constant:
```c++
const std::string asset_id = request.getArg ("id");
```

File: src/web/src/topology_input_power_chain.ecpp [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    if (asset_id.empty ()) {
+        std::string expected = TRANSLATE_ME("existing asset name");
+        http_die ("request-param-bad", "dc_id", asset_id.c_str (), expected.c_str ());
```
Comment (boricj) [inline src/web/src/topology_input_power_chain.ecpp]: Inline string construction:
```c++
http_die ("request-param-bad", "dc_id", asset_id.c_str (), TRANSLATE_ME("existing asset name").c_str ());
```

File: src/web/src/topology_input_power_chain.ecpp [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    zmsg_t *resp = client->requestreply (ADDRESS, SUBJECT, TIMEOUT, &req);
+    zmsg_destroy (&req);
+
+    #define CLEANUP { zmsg_destroy (&resp); }
+
+    if (!resp) {
+        CLEANUP;
+        log_error ("client->requestreply (timeout = %d s) failed", TIMEOUT);
+        std::string err = TRANSLATE_ME("Request to mlm client failed (timeout reached).");
+        http_die ("internal-error", err.c_str());
```
Comment (boricj) [inline src/web/src/topology_input_power_chain.ecpp]: Inline string construction:
```c++
http_die ("internal-error", TRANSLATE_ME("Request to mlm client failed (timeout reached).").c_str());
```

File: src/web/src/topology_location.ecpp [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+    const char *SUBJECT = "TOPOLOGY";
+    const char *COMMAND = "LOCATION";
+    const int TIMEOUT = 5; // seconds
+
+    // get params
+    std::string parameter_name;
+    std::string asset_id;
+    {
+        std::string from = qparam.param("from");
+        std::string to = qparam.param("to");
```
Comment (boricj) [inline src/web/src/topology_location.ecpp]: Make these variables constant.

File: src/web/src/topology_input_power_chain.ecpp (hunk 6 above)
Comment (jeffdegott) [reply src/web/src/topology_input_power_chain.ecpp]: http_die is a 'awfull-but-necessary' macro, as TRANSLATE_ME(). How many times TRANSLATE_ME would be instanciated by http_die()? Nobody knows... I prefer the current 'var' schema, personally. ;)

**Pull request merged**